flink-runtime/src/main/java/org/apache/flink/runtime/io/AvailabilityProvider.java

```diff
@@ -29,10 +29,30 @@
 public interface AvailabilityProvider {
 	/**
 	 * Constant that allows to avoid volatile checks {@link CompletableFuture#isDone()}. Check
-	 * {@link #getAvailableFuture()} for more explanation.
+	 * {@link #isAvailable()} and {@link #isVolatileAvailable()}for more explanation.
 	 */
 	CompletableFuture<?> AVAILABLE = CompletableFuture.completedFuture(null);
 
+	/**
+	 * Checks whether this instance is available via constant {@link #AVAILABLE} to avoid volatile access.
+	 *
+	 * @return true if this instance is available for further processing.
+	 */
+	default boolean isAvailable() {
```

Comment:
The main motivation of the 1st commit is actually from the 3rd commit. 

But there is another tiny motivation for it : my previous thought of `#isAvailable()` should return a boolean value exactly for better understanding, just like the way of `NetworkSequenceViewReader#isAvailable()` and `ResultSubpartitionView#isAvailable()`. If we do not bring the 3rd commit, it also makes sense to do the 1st commit from this point, so I make it separate.

Certainly I also accept to squash these two commit together and it can also be interpreted.

Reply:
I would squash them, otherwise the 1st commit was confusing me (please see the comments that I scratched out when I was doing the first pass of reviewing) and it also makes the reviewing more difficult (I would still like to look at those two changes combined, to see all of the `isAvailable()` -> `getAvailableFuture()` affected changes in one place/commit.

Reply:
I can squash them together.

> I guess with this commit, a few of these occassions should rather use the new is* functions.

Actually not that, the second commit would not touch the occasions of the first commit, because the current `isAvailable()` and `getAvailableFuture()` are motivated to use in different scenarios.

In the case of callback action, we use the `getAvailableFuture()` for monitoring the future.

In the case of only checking the current state, we use the `isAvailable()` to return boolean value which is currently used only in `StreamTwoInputProcessor` to capsulate the previous judging logics. In detail, currently we have two ways `available == AVAILABLE` and `available == AVAILABLE || available.isDone()` for checking this condition in different usages of performance/visibility concerns. After extracting these two ways into helper methods in `AvailabilityProvider` as this PR did, it has two benefits:

- Users can choose the way to use based on the specific scenarios, no need to implement this logic everywhere. I also tried to not expose constant `AvailabilityProvider#AVAILABLE` to the outside based on the provided methods, because I do not think it is easy to use properly before.

- If we plan to further improve the current usage of `CompleteFuture` in future, we could only focus on the internal implementation of these methods, no need to touch the usages in everywhere.